### PR TITLE
Push notifications

### PR DESCRIPTION
**Pull request by ifeins**
Server pull request for context: https://github.com/iic-ninjas/while-youre-at-it-server/pull/1

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -17,14 +17,26 @@ android {
       proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
     }
   }
+  productFlavors {
+    localFlavor {
+      //buildConfigField "String", "SERVER_HOST", "\"http://shopingo.192.168.2.217.xip.io/\""
+      buildConfigField "String", "SERVER_HOST", "\"http://25562db8.ngrok.com/\""
```
Comment (ifeins) [inline app/build.gradle]: will change to use powder xip address and the current host ip address

File: app/src/main/java/com/iic/shopingo/ShopingoApplication.java [hunk 2]
```diff
@@ -15,5 +14,11 @@ public void onCreate() {
     super.onCreate();
 
     CurrentUser.getInstance().setStorage(new UserStorage(PreferenceManager.getDefaultSharedPreferences(this)));
+    if (CurrentUser.getToken() != null) {
+      // TODO: needs to do this after login as well!!!!!
+      NotificationsHelper.registerForNotificationsAsync(this);
```
Comment (ifeins) [inline app/src/main/java/com/iic/shopingo/ShopingoApplication.java:18]: will do

File: app/src/main/java/com/iic/shopingo/services/GcmIntentService.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+    PendingIntent pendingIntent = null;
+    if (shopingoNotification instanceof TripNotification) {
+      TripNotification tripNotification = (TripNotification) shopingoNotification;
+      if (tripNotification.getStatus().equals("active")) {
+        Intent resultIntent = new Intent(this, SelectShopperActivity.class);
+        stackBuilder.addNextIntentWithParentStack(resultIntent);
+        pendingIntent = stackBuilder.getPendingIntent(0, PendingIntent.FLAG_UPDATE_CURRENT);
+      } else {
+        // TODO: implement cancelled
```
Comment (ifeins) [inline app/src/main/java/com/iic/shopingo/services/GcmIntentService.java]: working on it.

**Comment by assafgelber** [comment]
Looks good. I think we'll want to merge the UI thing branch before this so you can do the event bus things on the correct activity.

File: app/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -7,6 +7,14 @@
   <uses-permission android:name="android.permission.CALL_PHONE"/>
   <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION"/>
   <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION"/>
+  <uses-permission android:name="com.google.android.c2dm.permission.RECEIVE"/>
+  <uses-permission android:name="android.permission.WAKE_LOCK"/>
```
Comment (AsafGartner) [inline app/src/main/AndroidManifest.xml:11]: You probably don't need this.

File: app/src/main/AndroidManifest.xml [hunk 5]
```diff
@@ -84,7 +98,24 @@
...
+        >
+      <meta-data
+          android:name="android.support.PARENT_ACTIVITY"
+          android:value=".ui.HomeActivity"
+          />
+    </activity>
+
+    <receiver
+        android:name=".receivers.GcmBroadcastReceiver"
+        android:permission="com.google.android.c2dm.permission.SEND"
```
Comment (AsafGartner) [inline app/src/main/AndroidManifest.xml:110]: Do we need this?

File: app/src/main/java/com/iic/shopingo/services/GcmIntentService.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+  }
+
+  @Override
+  protected void onHandleIntent(Intent intent) {
+    Log.d(LOG_TAG, "Handling notification");
+    GoogleCloudMessaging gcm = GoogleCloudMessaging.getInstance(this);
+    String messageType = gcm.getMessageType(intent);
+
+    if (messageType.equals(GoogleCloudMessaging.MESSAGE_TYPE_MESSAGE)) {
+      String payload = intent.getStringExtra(EXTRA_PAYLOAD);
```
Comment (AsafGartner) [inline app/src/main/java/com/iic/shopingo/services/GcmIntentService.java:68]: GCM don't provide these constants?

Comment (ifeins) [inline app/src/main/java/com/iic/shopingo/services/GcmIntentService.java:68]: payload is something I defined.

File: app/src/main/java/com/iic/shopingo/ui/request_flow/activities/RequestStateActivity.java [hunk 7]
```diff
@@ -55,6 +58,34 @@ protected void onCreate(Bundle savedInstanceState) {
...
 
+  @Override
+  protected void onResume() {
+    super.onResume();
+    AppEventBus.getInstance().register(this);
+  }
+
+  @Override
+  protected void onPause() {
+    AppEventBus.getInstance().unregister(this);
```
Comment (AsafGartner) [inline app/src/main/java/com/iic/shopingo/ui/request_flow/activities/RequestStateActivity.java:66]: Move these to `onStart`/`onStop`.

File: app/src/main/java/com/iic/shopingo/ui/trip_flow/fragments/UnifiedShoppingListFragment.java [hunk 8]
```diff
@@ -54,6 +59,39 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
...
+  private void removeShoppingList(String requesterId) {
+    Iterator<ShoppingList> iterator = shoppingLists.iterator();
+    while (iterator.hasNext()) {
+      ShoppingList shoppingList = iterator.next();
+      if (shoppingList.requester.getId().equals(requesterId)) {
+        iterator.remove();
+      }
+    }
+
+    adapter.removeShoppingList(requesterId);
```
Comment (AsafGartner) [inline app/src/main/java/com/iic/shopingo/ui/trip_flow/fragments/UnifiedShoppingListFragment.java:91]: This should call `notifyDataSetChanged` on itself.

**Pull request merged**